README.md
# RFC4627 (JSON) and JSON-RPC for Erlang

An implementation of JSON and JSON-RPC for Erlang.

See
[rfc4627.erl](http://tonyg.github.com/erlang-rfc4627/doc/rfc4627.html),
the JSON/RFC4627 codec itself, to learn how to encode and decode JSON
objects from Erlang code.

## Providing and calling JSON-RPC services

See
[rfc4627\_jsonrpc.erl](http://tonyg.github.com/erlang-rfc4627/doc/rfc4627_jsonrpc.html),
a JSON-RPC service registry and transport-neutral service method
invocation mechanism, to learn how to expose Erlang processes as
remotely-callable JSON-RPC services, and to learn how to invoke local
JSON-RPC services from Erlang without the overhead of HTTP.

## Exposing JSON-RPC services over HTTP

### Using Inets

See
[rfc4627\_jsonrpc\_inets.erl](http://tonyg.github.com/erlang-rfc4627/doc/rfc4627_jsonrpc_inets.html),
an Inets HTTP transport binding for JSON-RPC, to learn how to
configure the Inets HTTP server to respond to JSON-RPC requests.

### Using Mochiweb

See
[rfc4627\_jsonrpc\_mochiweb.erl](http://tonyg.github.com/erlang-rfc4627/doc/rfc4627_jsonrpc_mochiweb.html)
to learn how to delegate incoming Mochiweb HTTP requests to the
JSON-RPC service dispatcher.

### Using Cowboy

See
[rfc4627\_jsonrpc\_cowboy.erl](http://tonyg.github.com/erlang-rfc4627/doc/rfc4627_jsonrpc_cowboy.html)
to learn how to delegate incoming Cowboy HTTP requests to the
JSON-RPC service dispatcher.

### Specifying as rebar dependency

To specify erlang-rfc4627 as dependency in your rebar config please use like below

```erlang
{deps, [
        {'erlang-rfc4627', ".*", {git, "git://github.com/tonyg/erlang-rfc4627.git", "master"}}
]}.
```

## Running the example test service that comes with the source code

Included with the Erlang RFC4627 source code is a small Inets-based
example that defines a "hello world"-style JSON-RPC service, and calls
it from a Javascript program embedded in a web page.

At your Erlang shell,

 - after compiling the code with `make all test-compile`,
 - when your current working directory contains the `test` directory
   from the distribution, such that
   `./test/server_root/conf/httpd.conf` exists,

type `test_jsonrpc_inets:start_httpd().` (Don't forget the trailing
"."!) This will

 - start the inets httpd on port 5671 (from `./test/server_root/conf/httpd.conf`)
 - allow HTTP access to JSON-RPC services via a url prefix of `/rpc` (again from `httpd.conf`)
 - start the `rfc4627_jsonrpc` service registry
 - register the test service

Visiting <http://localhost:5671/> in your browser should load a page
that uses javascript to invoke the Erlang-implemented JSON-RPC test
service.

## Invoking JSON-RPC procedures from Javascript

See the [relevant
section](http://tonyg.github.com/erlang-rfc4627/doc/overview-summary.html#Invoking_JSON-RPC_procedures_from_Javascript)
of the `edoc` documentation included with the source code.

## References

 - the [JSON RFC](http://www.ietf.org/rfc/rfc4627.txt).

 - the [JSON-RPC
   specification](http://json-rpc.org/wd/JSON-RPC-1-1-WD-20060807.html)
   1.1 working draft ([mirrored
   locally](http://tonyg.github.com/erlang-rfc4627/doc/JSON-RPC-1-1-WD-20060807.html)).

 - Joe Armstrong's
   [message](http://erlang.org/pipermail/erlang-questions/2005-November/017805.html)
   describing the basis of the JSON data type mapping that the
   `rfc4627.erl` module uses.

## Contributors

The codebase and documentation was originally written between 2007 and
2010 by Tony Garnock-Jones while at LShift in London. Since then, it
has been maintained by Tony and enhanced with a few generous
contributions from others:

 - András Veres-Szentkirályi
 - Eugene Volchek
 - Simon MacMullen
 - Andrey Khozov
 - Erik Timan

## Copyright and Licence

Copyrights:

 - Copyright &copy; 2007-2010, 2011, 2012 Tony Garnock-Jones.
 - Portions copyright 2007-2010 LShift Ltd.
 - Portions copyright contributors to the project, listed above.

License (MIT):

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
    BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
    ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
    CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.


debian/changelog
rfc4627-erlang (1.1.0-1) unstable; urgency=low

  * Unicode support per RFC.

 -- Tony Garnock-Jones <tonyg@shortstop.lshift.net>  Wed, 03 Oct 2007 15:58:39 +0100

rfc4627-erlang (1.0.4-1) unstable; urgency=low

  * Support X-JSON-RPC-Timeout header (value in milliseconds).

 -- Tony Garnock-Jones <tonyg@shortstop.lshift.net>  Thu, 21 Jun 2007 15:00:46 +0100

rfc4627-erlang (1.0.3-1) unstable; urgency=low

  * Add start_link for use in supervision hierarchies.

 -- Tony Garnock-Jones <tonyg@shortstop.lshift.net>  Wed, 20 Jun 2007 14:04:42 +0100

rfc4627-erlang (1.0.2-1) unstable; urgency=low

  * Documentation for release
  * Minor API improvements
  * MIT licensing

 -- Tony Garnock-Jones <tonyg@lshift.net>  Sat, 17 Feb 2007 14:44:57 +0000

rfc4627-erlang (1.0.1-1) unstable; urgency=low

  * JSON-RPC transport is factored out on the client
  * Header file path correction
  * No longer take any action synchronously in the JS client

 -- Tony Garnock-Jones <tonyg@shortstop.lshift.net>  Mon, 29 Jan 2007 15:28:06 +0000

rfc4627-erlang (1.0-1) unstable; urgency=low

  * Initial release

 -- Tony Garnock-Jones <tonyg@lshift.net>  Fri, 26 Jan 2007 14:17:16 +0000



src/mod_jsonrpc.erl
%% JSON-RPC for Erlang's inets httpd
%%---------------------------------------------------------------------------
%% Copyright (c) 2007 Tony Garnock-Jones <tonyg@kcbbs.gen.nz>
%% Copyright (c) 2007 LShift Ltd. <query@lshift.net>
%%
%% Permission is hereby granted, free of charge, to any person
%% obtaining a copy of this software and associated documentation
%% files (the "Software"), to deal in the Software without
%% restriction, including without limitation the rights to use, copy,
%% modify, merge, publish, distribute, sublicense, and/or sell copies
%% of the Software, and to permit persons to whom the Software is
%% furnished to do so, subject to the following conditions:
%%
%% The above copyright notice and this permission notice shall be
%% included in all copies or substantial portions of the Software.
%%
%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
%% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
%% MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
%% NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
%% BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
%% ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
%% CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%% SOFTWARE.
%%---------------------------------------------------------------------------

-module(mod_jsonrpc).
-include("rfc4627.hrl").
-include("mod_jsonrpc.hrl").
-include_lib("inets/src/httpd.hrl").

-behaviour(gen_server).

-export([start/0, start_link/0]).
-export([init/1, terminate/2, code_change/3, handle_call/3, handle_cast/2, handle_info/2]).
-export([do/1, load/2]).
-export([lookup_service/1, register_service/2, error_response/2, error_response/3, service/4, service/5, proc/2]).
-export([gen_object_name/0, service_address/2, system_describe/2]).
-export([jsonrpc_post/3, jsonrpc_post/4, invoke_service_method/6]).

start() ->
    gen_server:start({local, mod_jsonrpc}, ?MODULE, [], []).

start_link() ->
    gen_server:start_link({local, mod_jsonrpc}, ?MODULE, [], []).

do(ModData = #mod{data = OldData}) ->
    case {httpd_util:key1search(OldData, status),
	  httpd_util:key1search(OldData, response)} of
	{undefined, undefined} ->
	    do_rpc(ModData);
	_ ->
	    {proceed, OldData}
    end.

load("JsonRpcAlias " ++ Alias, []) ->
    {ok, [], {json_rpc_alias, Alias}}.

jsonrpc_post(ServiceRec, ModData, {obj, Fields}) ->
    jsonrpc_post(ServiceRec, ModData, {obj, Fields}, default).

jsonrpc_post(ServiceRec, ModData, {obj, Fields}, Timeout) ->
    Id = httpd_util:key1search(Fields, "id"),
    Method = httpd_util:key1search(Fields, "method"),
    Args = httpd_util:key1search(Fields, "params"),
    {Headers, ResultField} =
	expand_jsonrpc_reply(
	  invoke_service_method(ServiceRec, post, ModData, Method, Args, Timeout)),
    {Headers, build_jsonrpc_response(Id, ResultField)}.

build_jsonrpc_response(Id, ResultField) ->
    {obj, [{version, <<"1.1">>},
	   {id, Id},
	   ResultField]}.

lookup_service(Service) ->
    gen_server:call(mod_jsonrpc, {lookup_service, Service}).

do_rpc(ModData = #mod{data = OldData}) ->
    case extract_object_method_and_params(ModData) of
	no_match ->
	    {proceed, OldData};
	UriInfo = {_Object, _UriMethod, _UriParamStr} ->
	    {PostOrGet, Id, Service, Method, Args, Timeout} = parse_jsonrpc(ModData, UriInfo),
	    {Headers, ResultField} =
		expand_jsonrpc_reply(
		  case lookup_service(Service) of
		      not_found ->
			  error_response(404, "Service not found", Service);
		      ServiceRec ->
			  invoke_service_method(
			    ServiceRec, PostOrGet, ModData, Method, Args, Timeout)
		  end),
	    Result = build_jsonrpc_response(Id, ResultField),
	    ResultEnc = lists:flatten(rfc4627:encode(Result)),
	    {proceed, [{response, {response,
				   [{code, 200},
				    {content_length, integer_to_list(length(ResultEnc))},
				    {content_type, "text/plain"}%rfc4627:mime_type()}
				    | Headers],
				   ResultEnc}} | OldData]}
    end.

invoke_service_method(ServiceRec = #service{name = ServiceName},
		      PostOrGet, ModData, Method, Args, Timeout) ->
    case Method of
	<<"system.describe">> ->
	    {result, system_describe(service_address(ModData, ServiceName), ServiceRec)};
	<<"system.", _Rest/binary>> ->
	    error_response(403, "System methods forbidden", Method);
	_ ->
	    case lookup_service_proc(ServiceRec, Method) of
		{ok, ServiceProc} ->
		    invoke_service(PostOrGet, ServiceRec#service.handler,
				   ModData, ServiceProc, Args, Timeout);
		not_found ->
		    error_response(404, "Procedure not found", [ServiceName, Method])
	    end
    end.

expand_jsonrpc_reply(Reply = {result, _Value}) -> {[], Reply};
expand_jsonrpc_reply(Reply = {error, _Value}) -> {[], Reply};
expand_jsonrpc_reply({result, Value, Headers0}) -> {Headers0, {result, Value}};
expand_jsonrpc_reply({error, Value, Headers0}) -> {Headers0, {error, Value}}.

register_service(Pid, ServiceDescription) ->
    %%error_logger:info_msg("Registering ~p as ~p", [Pid, ServiceDescription]),
    gen_server:call(mod_jsonrpc, {register_service, Pid, ServiceDescription}).

error_response(Code, ErrorValue) when is_integer(Code) ->
    error_response(Code, "Error "++integer_to_list(Code), ErrorValue);
error_response(Message, ErrorValue) when is_list(Message) ->
    error_response(500, list_to_binary(Message), ErrorValue);
error_response(Message, ErrorValue) when is_binary(Message) ->
    error_response(500, Message, ErrorValue).

error_response(Code, Message, ErrorValue) when is_list(Message) ->
    error_response(Code, list_to_binary(Message), ErrorValue);
error_response(Code, Message, ErrorValue) ->
    {error, {obj, [{"name", <<"JSONRPCError">>},
		   {"code", Code},
		   {"message", Message},
		   {"error", ErrorValue}]}}.

service(Handler, Name, Id, Version, Procs) ->
    (service(Name, Id, Version, Procs))#service{handler = Handler}.

service(Name, Id, Version, Procs) when is_list(Name) ->
    service(list_to_binary(Name), Id, Version, Procs);
service(Name, Id, Version, Procs) when is_list(Id) ->
    service(Name, list_to_binary(Id), Version, Procs);
service(Name, Id, Version, Procs) when is_list(Version) ->
    service(Name, Id, list_to_binary(Version), Procs);
service(Name, Id, Version, Procs) ->
    #service{name = Name, id = Id, version = Version,
	     procs = lists:map(fun ({ProcName, Params}) ->
				       proc(ProcName, Params);
				   (P = #service_proc{}) ->
				       P
			       end, Procs)}.

proc(Name, Params) when is_list(Name) ->
    proc(list_to_binary(Name), Params);
proc(Name, Params) ->
    #service_proc{name = Name, params = lists:map(fun proc_param/1, Params)}.

proc_param({N, T}) when is_list(N) ->
    proc_param({list_to_binary(N), T});
proc_param({N, T}) ->
    #service_proc_param{name = N, type = proc_param_type(T)}.

proc_param_type(bit) -> <<"bit">>;
proc_param_type(num) -> <<"num">>;
proc_param_type(str) -> <<"str">>;
proc_param_type(arr) -> <<"arr">>;
proc_param_type(obj) -> <<"obj">>;
proc_param_type(any) -> <<"any">>;
proc_param_type(nil) -> <<"nil">>;
proc_param_type(T) when is_list(T) -> list_to_binary(T);
proc_param_type(T) when is_binary(T) -> T.

gen_object_name() ->
    Hash = erlang:md5(term_to_binary({node(), erlang:now()})),
    binary_to_hex(Hash).

binary_to_hex(<<>>) ->
    [];
binary_to_hex(<<B, Rest/binary>>) ->
    [rfc4627:hex_digit((B bsr 4) band 15),
     rfc4627:hex_digit(B band 15) |
     binary_to_hex(Rest)].

service_address(ModData, Object) when is_binary(Object) ->
    service_address(ModData, binary_to_list(Object));
service_address(#mod{socket_type = SocketType,
		     config_db = ConfigDb,
		     parsed_header = Headers},
		Object) ->
    AliasPrefix = httpd_util:lookup(ConfigDb, json_rpc_alias, "/jsonrpc"),
    Host = case httpd_util:key1search(Headers, "host") of
	       undefined -> "";
	       Name -> "//" ++ Name
	   end,
    Scheme = case SocketType of
		 ip_comm -> "http:";
		 ssl -> "https:";
		 _Other -> ""
	     end,
    Scheme ++ Host ++ AliasPrefix ++ "/" ++ Object;
service_address(_, Object) ->
    Object.

extract_object_method_and_params(#mod{config_db = ConfigDb, request_uri = Uri}) ->
    AliasPrefix = httpd_util:lookup(ConfigDb, json_rpc_alias, "/jsonrpc"),
    AliasPrefixRe = "^" ++ AliasPrefix ++ "/",
    case regexp:first_match(Uri, AliasPrefixRe) of
	{match, 1, Length} ->
	    Tail = string:substr(Uri, Length + 1),
	    {ObjectMethod, Params} = case string:tokens(Tail, "?") of
					 [] -> {"", ""};
					 [Method] -> {Method, ""};
					 [Method, QueryStr] -> {Method, QueryStr}
				     end,
	    case lists:reverse(string:tokens(ObjectMethod, "/")) of
		[] -> {"", "", Params};
		[Object] -> {list_to_binary(Object), <<>>, Params};
		[Method1, Object | _] -> {list_to_binary(Object), list_to_binary(Method1), Params}
	    end;
	nomatch ->
	    no_match
    end.

extract_timeout_header(#mod{parsed_header = ParsedHeader}) ->
    case httpd_util:key1search(ParsedHeader, "x-json-rpc-timeout", "default") of
	"default" ->
	    default;
	"infinity" ->
	    infinity;
	NumStr ->
	    list_to_integer(NumStr)
    end.

parse_jsonrpc(ModData = #mod{method = "POST", entity_body = Body},
	      {Object, _UriMethod, _UriParams}) ->
    {ok, {obj, Fields}, _} = rfc4627:decode(Body),
    {post,
     httpd_util:key1search(Fields, "id"),
     Object,
     httpd_util:key1search(Fields, "method"),
     httpd_util:key1search(Fields, "params"),
     extract_timeout_header(ModData)};
parse_jsonrpc(ModData = #mod{method = _},
	     {Object, Method, QueryStr}) ->
    %% GET, presumably. We don't really care, here.
    Q = case httpd:parse_query(QueryStr) of
	    [{"",""}] -> []; %% is this a bug in httpd?
	    ActualParameters -> ActualParameters
	end,
    {get,
     null,
     Object,
     Method,
     %% FIXME: need to collect duplicate parameter keys into a list, as per spec
     {obj, lists:map(fun ({Key, ValStr}) ->
			     {Key, list_to_binary(ValStr)}
		     end, Q)},
     extract_timeout_header(ModData)}.

lookup_service_proc(#service{procs = Procs}, Method) ->
    case lists:keysearch(Method, #service_proc.name, Procs) of
	{value, ServiceProc} ->
	    {ok, ServiceProc};
	false ->
	    not_found
    end.

invoke_service(get, Handler, ModData, ServiceProc, Args, Timeout) ->
    if
	ServiceProc#service_proc.idempotent ->
	    invoke_service1(Handler, ModData, ServiceProc, Args, Timeout);
	true ->
	    error_response(403, "Non-idempotent method", ServiceProc#service_proc.name)
    end;
invoke_service(post, Handler, ModData, ServiceProc, Args, Timeout) ->
    invoke_service1(Handler, ModData, ServiceProc, Args, Timeout).

invoke_service1(Handler, ModData, #service_proc{name = Name, params = Params}, Args, Timeout) ->
    %%error_logger:info_msg("JSONRPC invoking ~p:~p(~p)", [Handler, Name, Args]),
    case catch run_handler(Handler, Name, ModData, coerce_args(Params, Args), Timeout) of
	{'EXIT', {{function_clause, _}, _}} ->
	    error_response(404, "Undefined procedure", Name);
	{'EXIT', Reason} ->
	    error_response(500, "Internal error", list_to_binary(io_lib:format("~p", [Reason])));
	Response ->
	    Response
    end.

run_handler({pid, Pid}, Name, ModData, CoercedArgs, default) ->
    gen_server:call(Pid, {jsonrpc, Name, ModData, CoercedArgs});
run_handler({pid, Pid}, Name, ModData, CoercedArgs, Timeout) ->
    gen_server:call(Pid, {jsonrpc, Name, ModData, CoercedArgs}, Timeout);
run_handler({function, F}, Name, ModData, CoercedArgs, _Timeout) ->
    F(Name, ModData, CoercedArgs).

coerce_args(_Params, Args) when is_list(Args) ->
    Args;
coerce_args(Params, {obj, Fields}) ->
    lists:map(fun (#service_proc_param{name = Name, type = Type}) ->
		      case lists:keysearch(binary_to_list(Name), 1, Fields) of
			  {value, {_, Value}} ->
			      coerce_value(Value, Type);
			  false ->
			      null
		      end
	      end, Params).

coerce_value(Value, _Type) when not(is_binary(Value)) ->
    Value;
coerce_value(<<"true">>, <<"bit">>) -> true;
coerce_value(_, <<"bit">>) -> false;
coerce_value(V, <<"num">>) -> list_to_integer(binary_to_list(V));
coerce_value(V, <<"str">>) -> V;
coerce_value(V, <<"arr">>) -> rfc4627:decode(V);
coerce_value(V, <<"obj">>) -> rfc4627:decode(V);
coerce_value(V, <<"any">>) -> V;
coerce_value(_, <<"nil">>) -> null;
coerce_value(V, _) -> V.

remove_undefined({obj, Fields}) ->
    {obj, remove_undefined1(Fields)}.

remove_undefined1([]) ->
    [];
remove_undefined1([{_, undefined} | Rest]) ->
    remove_undefined1(Rest);
remove_undefined1([X | Rest]) ->
    [X | remove_undefined1(Rest)].

system_describe(AddressStr,
		#service{name = Name, id = Id, version = Version, summary = Summary,
			 help = Help, procs = Procs}) ->
    remove_undefined({obj, [{"sdversion", <<"1.0">>},
			    {"name", Name},
			    {"id", Id},
			    {"version", Version},
			    {"summary", Summary},
			    {"help", Help},
			    {"address", list_to_binary(AddressStr)},
			    {"procs", lists:map(fun system_describe_proc/1, Procs)}]}).

system_describe_proc(P = #service_proc{params = Params}) ->
    remove_undefined(?RFC4627_FROM_RECORD(service_proc,
					  P#service_proc{params =
							 lists:map(fun system_describe_proc_param/1,
								   Params)})).

system_describe_proc_param(P = #service_proc_param{}) ->
    remove_undefined(?RFC4627_FROM_RECORD(service_proc_param, P)).

%---------------------------------------------------------------------------

init(_Args) ->
    {ok, no_jsonrpc_state}.

terminate(_Reason, _State) ->
    %% FIXME: should we notify services here?
    ok.

code_change(_OldVsn, State, _Extra) ->
    State.

handle_call({lookup_service, Service}, _From, State) ->
    case get({service, Service}) of
	undefined ->
	    {reply, not_found, State};
	ServiceRec ->
	    {reply, ServiceRec, State}
    end;

handle_call({register_service, Pid, ServiceDescription}, _From, State) ->
    SD = ServiceDescription#service{handler = {pid, Pid}},
    erlang:monitor(process, Pid),
    put({service_pid, Pid}, SD#service.name),
    put({service, SD#service.name}, SD),
    {reply, ok, State}.

handle_cast(Request, State) ->
    error_logger:error_msg("Unhandled cast in mod_jsonrpc: ~p", [Request]),
    {noreply, State}.

handle_info({'DOWN', _MonitorRef, process, DownPid, _Reason}, State) ->
    case get({service_pid, DownPid}) of
	undefined ->
	    %% How strange.
	    {noreply, State};
	ServiceName ->
	    erase({service_pid, DownPid}),
	    erase({service, ServiceName}),
	    {noreply, State}
    end.
